common/workunit/workunit.cpp

```diff
@@ -6805,6 +6837,122 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
     throwUnexpected();
 }
 
+bool extractFromWorkunitDAToken(const char * distributedAccessToken, StringBuffer * wuid, StringBuffer * user, StringBuffer * privKey)
+{
+    if (!isWorkunitDAToken(distributedAccessToken))
+    {
+        DBGLOG("Not a valid workunit distributed access token");
+        return false;
+    }
+
+    //Extract the string between [ and ]
+    const char * pEndBracket = strchr(distributedAccessToken + 5, ']');
+    StringBuffer tokenValues;
+    tokenValues.append(pEndBracket - distributedAccessToken - 5, distributedAccessToken + 5);
+
+    StringArray nameValues;
+    nameValues.appendList(tokenValues.str(), ",",true);
+
+    if (user || wuid)
+    {
+        for (int x = 0; x < nameValues.ordinality(); x++)
+        {
+            if (user && 0==strncmp(nameValues[x],"u=",2))//Extract user
+                user->append(nameValues[x] + 2);
+            if (wuid && 0==strncmp(nameValues[x],"w=",2))//Extract wuid
+                wuid->append(nameValues[x] + 2);
+        }
+    }
+
+    if (privKey)
+    {
+        const char * finger = pEndBracket;
+        if (finger)
+        {
+            ++finger;
+            while (*finger && *finger != ';')
+                privKey->append(1, (const char *)finger++);
+        }
+    }
+    return true;
+}
+
+//Verifies the given signed workunit distributed access token was created
+//and signed by the given wuid and user, and that the workunit is still active
+//Returns:
+//   0 : Success, token is valid and workunit is active
+//   1 : Signature does not verify (wuid/username don't match, or signature doesnt verify)
+//   2 : Workunit not active
+//   Throws if unable to open workunit
+CriticalSection verifyWorkunitDATokenCrit;
+int verifyWorkunitDAToken(const char * distributedAccessToken)
+{
+    CriticalBlock block(verifyWorkunitDATokenCrit);
+    StringBuffer tokWuid;
+    StringBuffer tokUser;
+    if (!extractFromWorkunitDAToken(distributedAccessToken, &tokWuid, &tokUser, nullptr))//get the wuid and user
+    {
+        //Not a valid workunit distributed access token
+        return false;
+    }
+
+    //Validate signature
+    IDigitalSignatureManager * pDSM = queryDigitalSignatureManagerInstanceFromEnv();
+    if (pDSM && pDSM->isDigiVerifierConfigured())
+    {
+        const char * finger;
+        StringBuffer token;//receives copy of everything up until signature
+        for (finger = distributedAccessToken; *finger && *finger != ';'; finger++)
+            token.append(1, finger);
+        token.append(1, finger);//append ;
+
+        if (finger && *finger)
+        {
+            StringBuffer sig(++finger);
+            if (!pDSM->digiVerify(token, sig))
```

Comment:
Maybe the same as above, but is there a JIRA open about digiVerify being told which public key to use for this particular signatue?